File: double_slit_simulation.py
import numpy as np
import matplotlib.pyplot as plt

# Constants and parameters
PI = np.pi
WAVELENGTH = 0.5
SLIT_DISTANCE = 1.0
SCREEN_DISTANCE = 10.0
NUM_POINTS = 1000
INTENSITY_SCALE = 1000.0

def calculate_intensity(x):
    path_diff1 = np.sqrt((x - SLIT_DISTANCE / 2) ** 2 + SCREEN_DISTANCE ** 2)
    path_diff2 = np.sqrt((x + SLIT_DISTANCE / 2) ** 2 + SCREEN_DISTANCE ** 2)

    phase_diff = 2 * PI * (path_diff1 - path_diff2) / WAVELENGTH

    amplitude = 2.0 * np.cos(phase_diff / 2.0)
    intensity = amplitude ** 2

    return intensity

def simulate_experiment():
    screen_range = 20.0
    step_size = screen_range / NUM_POINTS
    x_values = np.linspace(-screen_range / 2, screen_range / 2, NUM_POINTS)
    intensities = np.array([calculate_intensity(x) * INTENSITY_SCALE for x in x_values])

    # Save the data to a file
    with open("double_slit_simulation.txt", "w") as file:
        for x, intensity in zip(x_values, intensities):
            file.write(f"{x} {intensity}\n")

    # Plot the results
    plt.plot(x_values, intensities)
    plt.xlabel("Position on Screen")
    plt.ylabel("Intensity")
    plt.title("Double-Slit Experiment Simulation")
    plt.show()

if __name__ == "__main__":
    simulate_experiment()

File: double_slit_simulation.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>


#define PI 3.141592653589793


double calculateIntensity(double wavelength, double slitDistance, double screenDistance, double x) {
    double k = (2 * PI) / wavelength;
    double d = slitDistance; 
    double z = screenDistance; 
    double pathDifference = (d * x) / z;
    double phaseDifference = k * pathDifference;
    double intensity = pow(cos(phaseDifference / 2), 2);
    return intensity;
}

int main() {
    double wavelength, slitDistance, screenDistance, x;

    // User inputs
    printf("Enter the wavelength (in meters): ");
    scanf("%lf", &wavelength);

    printf("Enter the slit distance (in meters): ");
    scanf("%lf", &slitDistance);

    printf("Enter the screen distance (in meters): ");
    scanf("%lf", &screenDistance);

    
    printf("\nPosition on Screen (m)\tIntensity\n");
    printf("----------------------\t---------\n");

    
    for (x = -0.01; x <= 0.01; x += 0.0001) {
        double intensity = calculateIntensity(wavelength, slitDistance, screenDistance, x);
        printf("%-22.4e\t%-10.4e\n", x, intensity);
    }

    return 0;
}
